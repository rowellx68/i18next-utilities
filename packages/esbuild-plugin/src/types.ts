import { IncludePattern } from 'i18next-utilities-core';

export type Options = {
  /**
   * Glob pattern to include files for bundling. Defaults to `['**\/*.json', '**\/*.yml', '**\/*.yaml']`.
   */
  include?: IncludePattern[];

  /**
   * Namespace resolution strategy. Defaults to `'none'`.
   */
  namespaceResolution?: 'basename' | 'relativePath';

  /**
   * Locale top-level directory paths. Ordered from least to most specific.
   */
  paths: string[];
};
