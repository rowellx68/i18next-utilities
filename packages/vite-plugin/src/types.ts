import type { LogLevel } from 'vite';

export type IncludePattern = '**/*.json' | '**/*.yml' | '**/*.yaml';

export type I18NextTypedDtsOptions = {
  expand: {
    arrays: boolean;
    contexts: boolean;
    ordinals: boolean;
    plurals: boolean;
  };
};

export type I18NextTypedLoaderOptions = {
  /**
   * Log level.
   * @default 'warn'
   */
  logLevel?: LogLevel;

  /**
   * Glob pattern to include files for bundling.
   * @default ['**\/*.json', '**\/*.yml', '**\/*.yaml']
   */
  include?: IncludePattern[];

  /**
   * Namespace resolution strategy.
   */
  namespaceResolution?: 'basename' | 'relativePath';

  /**
   * Default namespace for files that do not have a namespace specified.
   * @default 'translation'
   */
  defaultNamespace?: string;

  /**
   * Default locale which will be used to generate the types.
   * @default 'en'
   */
  defaultLocale?: string;

  /**
   * Locale top-level directory paths. Ordered from least to most specific.
   */
  paths: string[];

  /**
   * Output file destination for the generated types.
   * @default './src/types/i18next.d.ts'
   */
  dtsOutputFile?: string;

  /**
   * Output file destination for the generated virtual module types.
   * @default './src/types/i18next-virtual.d.ts'
   */
  virtualModuleDtsOutputFile?: string;

  /**
   * Options for generating the type definitions.
   *
   * When set to `true`, the arrays/ordinals/plurals will be expanded.
   *
   * @default { expand: { arrays: false, ordinals: false, plurals: false } }
   */
  dts?: I18NextTypedDtsOptions;
};
