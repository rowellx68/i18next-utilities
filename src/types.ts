import type { LogLevel } from 'vite';

export type IncludePattern = '**/*.json' | '**/*.yml' | '**/*.yaml';

export type ResourceBundle = { [key: string]: string | object };

export type I18NextTypedLoaderOptions = {
  /**
   * Log level.
   * @default 'warn'
   */
  logLevel?: LogLevel;

  /**
   * Glob pattern to include files for bundling.
   * @default ['**\/*.json', '**\/*.yml', '**\/*.yaml']
   */
  include?: IncludePattern[];

  /**
   * Namespace resolution strategy.
   */
  namespaceResolution?: 'basename' | 'relativePath';

  /**
   * Default namespace for files that do not have a namespace specified.
   * @default 'translation'
   */
  defaultNamespace?: string;

  /**
   * Default locale which will be used to generate the types.
   * @default 'en'
   */
  defaultLocale?: string;

  /**
   * Locale top-level directory paths. Ordered from least to most specific.
   */
  paths: string[];

  /**
   * Output file destination for the generated types.
   * @default './src/types/i18next.d.ts'
   */
  dtsOutputFile?: string;
};
