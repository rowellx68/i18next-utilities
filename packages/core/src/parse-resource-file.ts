import path from 'path';
import fs from 'fs';
import YAML from 'yaml';
import * as deep from 'ts-deepmerge';
import * as prop from 'dot-prop';
import type { Resource } from 'i18next';
import type {
  I18NextTypedLogger,
  I18NextTypedOptions,
  IncludePattern,
} from './types';
import {
  resolvePaths,
  assertExistence,
  enumerateLanguages,
  findAllFiles,
} from './utilities';

const defaultIncludes: IncludePattern[] = [
  '**/*.json',
  '**/*.yml',
  '**/*.yaml',
];

const allowedExtensions = ['.json', '.yml', '.yaml'];

export const parseResourceFiles = (
  options: I18NextTypedOptions,
  logger: I18NextTypedLogger
) => {
  const directories = resolvePaths(options.paths, process.cwd());
  const files: string[] = [];

  assertExistence(directories);

  const uniqueIncludes = Array.from(
    new Set(options.include ?? defaultIncludes)
  );

  if (options.paths.length === 0) {
    logger.warn('No paths to search for files.');

    return { files, bundle: {}, defaultBundle: {} };
  }

  if (uniqueIncludes.length === 0) {
    logger.warn('No includes patterns specified.');

    return { files, bundle: {}, defaultBundle: {} };
  }

  let allLanguages: Set<string> = new Set();
  let appResourceBundle: Resource = {};

  for (const directory of directories) {
    const languages = enumerateLanguages(directory);
    allLanguages = new Set([...allLanguages, ...languages]);

    for (const language of languages) {
      const resourceBundle: Resource = {};
      resourceBundle[language] = {};

      const languageDirectory = path.join(directory, language);
      const resourceFiles = findAllFiles(uniqueIncludes, languageDirectory);

      for (const file of resourceFiles) {
        const fullPath = path.resolve(directory, language, file);
        const extension = path.extname(file);

        if (!allowedExtensions.includes(extension)) {
          logger.warn(`Unsupported file: ${file}`);
          continue;
        }

        files.push(fullPath);

        const fileContent = fs.readFileSync(fullPath, 'utf8');

        const content =
          extension === '.json'
            ? JSON.parse(String(fileContent))
            : YAML.parse(String(fileContent));

        if (options.namespaceResolution) {
          let namespaceFilePath = file;

          if (options.namespaceResolution === 'basename') {
            namespaceFilePath = path.basename(file);
          } else if (options.namespaceResolution === 'relativePath') {
            namespaceFilePath = path.relative(
              path.join(directory, language),
              file
            );
          }

          const nsparts = namespaceFilePath
            .replace(extension, '')
            .split(path.sep)
            .filter((part) => part !== '' && part !== '..');

          const namespace = [language].concat(nsparts).join('.');

          prop.setProperty(resourceBundle, namespace, content);
        } else {
          resourceBundle[language] = content;
        }

        appResourceBundle = deep.merge(appResourceBundle, resourceBundle);
      }
    }
  }

  const defaultBundle = appResourceBundle[options.defaultLocale || 'en'];

  return { files, bundle: appResourceBundle, defaultBundle };
};
