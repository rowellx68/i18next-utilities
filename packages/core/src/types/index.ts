export type IncludePattern = '**/*.json' | '**/*.yml' | '**/*.yaml';

export type I18NextTypedDtsOptions = {
  expand: boolean;
};

export type I18NextTypedOptions = {
  /**
   * Glob pattern to include files for bundling.
   * @default ['**\/*.json', '**\/*.yml', '**\/*.yaml']
   */
  include?: IncludePattern[];

  /**
   * Namespace resolution strategy.
   */
  namespaceResolution?: 'basename' | 'relativePath';

  /**
   * Default namespace for files that do not have a namespace specified.
   * @default 'translation'
   */
  defaultNamespace?: string;

  /**
   * Default locale which will be used to generate the types.
   * @default 'en'
   */
  defaultLocale?: string;

  /**
   * Locale top-level directory paths. Ordered from least to most specific.
   */
  paths: string[];

  /**
   * Options for generating the type definitions.
   *
   * When set to `true`, the arrays/ordinals/plurals will be expanded.
   *
   * @default { expand: true }
   */
  dts?: I18NextTypedDtsOptions;
};

export type I18NextTypedLogger = {
  info: (message: string) => void;
  warn: (message: string) => void;
  error: (message: string) => void;
};
