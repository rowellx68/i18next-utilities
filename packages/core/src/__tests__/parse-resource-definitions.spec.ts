import { describe, it, expect, vi } from 'vitest';
import { parseResourceFiles } from '../parse-resource-file';
import { IncludePattern } from '../types';

describe('parseResourceFiles', () => {
  it.each([undefined, 'basename', 'relativePath'] as (
    | undefined
    | 'basename'
    | 'relativePath'
  )[])(
    'should parse all resource files with %s ns resolution',
    (namespaceResolution) => {
      const logger = {
        warn: vi.fn(),
        error: vi.fn(),
        info: vi.fn(),
      };

      const { files, defaultBundle, bundle } = parseResourceFiles(
        {
          paths: ['./src/__tests__/__fixtures__/locales'],
          namespaceResolution: namespaceResolution,
        },
        logger
      );

      expect(bundle).toMatchSnapshot();
      expect(defaultBundle).toMatchSnapshot();
      expect(files).toHaveLength(3);
    }
  );

  it.each<IncludePattern>(['**/*.json', '**/*.yaml', '**/*.yml'])(
    'should parse resource files for pattern %s',
    (pattern) => {
      const logger = {
        warn: vi.fn(),
        error: vi.fn(),
        info: vi.fn(),
      };

      const { files, defaultBundle, bundle } = parseResourceFiles(
        {
          paths: ['./src/__tests__/__fixtures__/locales'],
          namespaceResolution: 'basename',
          include: [pattern],
        },
        logger
      );

      expect(bundle).toMatchSnapshot();
      expect(defaultBundle).toMatchSnapshot();
      expect(files).toHaveLength(1);
    }
  );

  it('calls logger.warn when no paths are included', () => {
    const logger = {
      warn: vi.fn(),
      error: vi.fn(),
      info: vi.fn(),
    };

    const { files, defaultBundle, bundle } = parseResourceFiles(
      {
        paths: [],
        namespaceResolution: 'basename',
      },
      logger
    );

    expect(logger.warn).toHaveBeenCalledWith('No paths to search for files.');
    expect(files).toHaveLength(0);
    expect(bundle).toEqual({});
    expect(defaultBundle).toEqual({});
  });

  it('calls logger.warn when no includes are specified', () => {
    const logger = {
      warn: vi.fn(),
      error: vi.fn(),
      info: vi.fn(),
    };

    const { files, defaultBundle, bundle } = parseResourceFiles(
      {
        paths: ['./src/__tests__/__fixtures__/locales'],
        namespaceResolution: 'basename',
        include: [],
      },
      logger
    );

    expect(logger.warn).toHaveBeenCalledWith('No includes patterns specified.');
    expect(files).toHaveLength(0);
    expect(bundle).toEqual({});
    expect(defaultBundle).toEqual({});
  });
});
