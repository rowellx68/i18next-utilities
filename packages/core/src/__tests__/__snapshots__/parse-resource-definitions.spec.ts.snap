// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseResourceFiles > should parse all resource files with basename ns resolution 1`] = `
{
  "en": {
    "translations": {
      "app": {
        "name": "My App",
      },
    },
  },
  "en-GB": {
    "other": {
      "dashboard": {
        "title": "Dashboard",
      },
    },
    "translations": {
      "app": {
        "description": "This is a test app",
      },
    },
  },
}
`;

exports[`parseResourceFiles > should parse all resource files with basename ns resolution 2`] = `
{
  "translations": {
    "app": {
      "name": "My App",
    },
  },
}
`;

exports[`parseResourceFiles > should parse all resource files with relativePath ns resolution 1`] = `
{
  "en": {
    "translations": {
      "app": {
        "name": "My App",
      },
    },
  },
  "en-GB": {
    "other": {
      "dashboard": {
        "title": "Dashboard",
      },
    },
    "translations": {
      "app": {
        "description": "This is a test app",
      },
    },
  },
}
`;

exports[`parseResourceFiles > should parse all resource files with relativePath ns resolution 2`] = `
{
  "translations": {
    "app": {
      "name": "My App",
    },
  },
}
`;

exports[`parseResourceFiles > should parse all resource files with undefined ns resolution 1`] = `
{
  "en": {
    "app": {
      "name": "My App",
    },
  },
  "en-GB": {
    "app": {
      "description": "This is a test app",
    },
    "dashboard": {
      "title": "Dashboard",
    },
  },
}
`;

exports[`parseResourceFiles > should parse all resource files with undefined ns resolution 2`] = `
{
  "app": {
    "name": "My App",
  },
}
`;

exports[`parseResourceFiles > should parse resource files for pattern **/*.json 1`] = `
{
  "en": {
    "translations": {
      "app": {
        "name": "My App",
      },
    },
  },
}
`;

exports[`parseResourceFiles > should parse resource files for pattern **/*.json 2`] = `
{
  "translations": {
    "app": {
      "name": "My App",
    },
  },
}
`;

exports[`parseResourceFiles > should parse resource files for pattern **/*.yaml 1`] = `
{
  "en-GB": {
    "other": {
      "dashboard": {
        "title": "Dashboard",
      },
    },
  },
}
`;

exports[`parseResourceFiles > should parse resource files for pattern **/*.yaml 2`] = `undefined`;

exports[`parseResourceFiles > should parse resource files for pattern **/*.yml 1`] = `
{
  "en-GB": {
    "translations": {
      "app": {
        "description": "This is a test app",
      },
    },
  },
}
`;

exports[`parseResourceFiles > should parse resource files for pattern **/*.yml 2`] = `undefined`;
